s3-source-connector/src/main/java/io/aiven/kafka/connect/s3/source/utils/FileReader.java

```diff
@@ -51,36 +51,60 @@ public FileReader(final S3SourceConfig s3SourceConfig, final String bucketName,
     }
 
     @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
-    List<S3ObjectSummary> fetchObjectSummaries(final AmazonS3 s3Client) throws IOException {
-        final List<S3ObjectSummary> allSummaries = new ArrayList<>();
-        String continuationToken = null;
-        ListObjectsV2Result objectListing;
-
-        do {
-            // Create the request for listing objects
-            final ListObjectsV2Request request = new ListObjectsV2Request().withBucketName(bucketName)
-                    .withMaxKeys(s3SourceConfig.getInt(FETCH_PAGE_SIZE) * PAGE_SIZE_FACTOR)
-                    .withContinuationToken(continuationToken); // Set continuation token for pagination
-
-            // List objects from S3
-            objectListing = s3Client.listObjectsV2(request);
-
-            // Filter out zero-byte objects and add to the list
-            final List<S3ObjectSummary> filteredSummaries = objectListing.getObjectSummaries()
-                    .stream()
-                    .filter(objectSummary -> objectSummary.getSize() > 0)
-                    .filter(objectSummary -> !failedObjectKeys.contains(objectSummary.getKey()))
-                    .collect(Collectors.toList());
-
-            allSummaries.addAll(filteredSummaries); // Add the filtered summaries to the main list
-
-            allSummaries.forEach(objSummary -> LOGGER.info("Objects to be processed {} ", objSummary.getKey()));
-
-            // Check if there are more objects to fetch
-            continuationToken = objectListing.getNextContinuationToken();
-        } while (objectListing.isTruncated()); // Continue fetching if the result is truncated
-
-        return allSummaries;
+    Iterator<S3ObjectSummary> fetchObjectSummaries(final AmazonS3 s3Client) {
+        return new Iterator<>() {
+            private String continuationToken = null; // NOPMD
+            private List<S3ObjectSummary> currentBatch = new ArrayList<>();
+            private int currentIndex = 0; // NOPMD
+            private boolean isTruncated = true;
+
+            @Override
+            public boolean hasNext() {
+                // If there are unprocessed objects in the current batch, we return true
+                if (currentIndex < currentBatch.size()) {
+                    return true;
+                }
+
+                if (isTruncated) {
+                    fetchNextBatch();
+                    return !currentBatch.isEmpty();
+                }
+
+                return false;
+            }
+
+            @Override
+            public S3ObjectSummary next() {
+                if (!hasNext()) {
+                    return null;
```

Comment:
Does this violates the contract for Iterator? The javadoc says that if there are no element an exception is thrown

Reply:
No, if exception is thrown, it would shutdown the connector if not caught again. I have a check to check for null in main iterator.

Reply:
We can alternatively return an optional

Reply:
Will be handled in ticket https://aiven.atlassian.net/browse/KCON-33